benchexec/resources.py

```diff
@@ -57,149 +63,332 @@ def get_cpu_cores_per_run(
-        unused_cores = []
-        for unit, cores in cores_of_unit.items():
-            for core in cores:
-                unit_of_core[core] = unit
-        for core, siblings in siblings_of_core.items():
-            if core in allCpus:
-                siblings.remove(core)
-                cores_of_unit[unit_of_core[core]] = [
-                    c for c in cores_of_unit[unit_of_core[core]] if c not in siblings
-                ]
-                siblings_of_core[core] = [core]
-                allCpus = [c for c in allCpus if c not in siblings]
-            else:
-                unused_cores.append(core)
-        for core in unused_cores:
-            siblings_of_core.pop(core)
-        logging.debug(
-            "Running in no-hyperthreading mode, avoiding the use of CPU cores %s",
-            unused_cores,
-        )
+    coreLimit_rounded_up = calculate_coreLimit_rounded_up(siblings_of_core, coreLimit)
+    chosen_level = calculate_chosen_level(hierarchy_levels, coreLimit_rounded_up)
 
-    unit_size = len(next(iter(cores_of_unit.values())))  # Number of units per core
-    if any(len(cores) != unit_size for cores in cores_of_unit.values()):
-        sys.exit(
-            "Asymmetric machine architecture not supported: "
-            "CPUs/memory regions with different number of cores."
-        )
+    # calculate runs per unit of hierarchy level i
+    unit_size = len(next(iter(hierarchy_levels[chosen_level].values())))
+    assert unit_size >= coreLimit_rounded_up
+    runs_per_unit = int(math.floor(unit_size / coreLimit_rounded_up))
 
-    core_size = len(next(iter(siblings_of_core.values())))  # Number of threads per core
-    if any(len(siblings) != core_size for siblings in siblings_of_core.values()):
-        sys.exit(
-            "Asymmetric machine architecture not supported: "
-            "CPU cores with different number of sibling cores."
-        )
+    # compare num of units & runs per unit vs num_of_threads
+    if len(hierarchy_levels[chosen_level]) * runs_per_unit < num_of_threads:
+        if not isTest:
+            sys.exit(
+                f"Cannot assign required number of threads."
+                f"Please reduce the number of threads to {len(hierarchy_levels[chosen_level]) * runs_per_unit}."
+            )
+        else:
+            is_feasible = False
+
+    # calculate if sub_units have to be split to accommodate the runs_per_unit
+    sub_units_per_run = calculate_sub_units_per_run(
+        coreLimit_rounded_up, hierarchy_levels, chosen_level
+    )
+    print("sub_units_per_run = ", sub_units_per_run)
+    # number of nodes at subunit-Level / sub_units_per_run
+    if len(hierarchy_levels[chosen_level - 1]) / sub_units_per_run < num_of_threads:
+        if not isTest:
+            sys.exit(
+                f"Cannot split memory regions between runs. "
+                f"Please reduce the number of threads to {math.floor(len(hierarchy_levels[chosen_level-1]) / sub_units_per_run)}."
```

Comment:
For long expressions like this (that exceed the line length) I guess it is better readable if the expression is stored in an additional variable first.

Reply:
Done in commit https://github.com/sosy-lab/benchexec/pull/892/commits/0689df517323f46784726a9fc5b3d74c0abcdad0